api/src/sparql.rs

```diff
@@ -0,0 +1,104 @@
+//! Common traits for working with [SPARQL](https://www.w3.org/TR/sparql11-query/).
+//!
+//! # Design rationale
+//!
+//! These traits are deliberately very generic.
+//! Specific implementations may have additional features, such as:
+//!
+//! - preparing a query for multiple use;
+//! - setting default values for `BASE`, `PREFIX`, `FROM`, `FROM NAMED` directives,
+//!   before parsing query string;
+//! - pre-binding variables before evaluating query;
+//! - etc...
+//!
+//! However, we do not want to impose these feature, or any subset thereof,
+//! to all implementation of Sophia.
+
+use crate::dataset::Dataset;
+use crate::graph::Graph;
+use crate::term::TTerm;
+use std::error::Error;
+
+/// A dataset that can be queried with SPARQL.
+pub trait SparqlDataset: Dataset {
+    type BindingsTerm: TTerm;
+    type BindingsResult: SparqlBindings<Self::BindingsTerm>;
+    type GraphResult: Graph;
+    type SparqlError: Error + 'static;
+
+    /// Parse and immediately execute `query`
+    fn query(&self, query: &str) -> Result<SparqlResult<Self>, Self::SparqlError>;
+}
+
+/// The result of executing a SPARQL query.
+pub enum SparqlResult<T>
+where
+    T: SparqlDataset + ?Sized,
+{
+    Bindings(T::BindingsResult),
+    Boolean(bool),
+    Graph(T::GraphResult),
+}
+
+/// The result of executing a SPARQL SELECT query
+pub trait SparqlBindings<T: TTerm>: IntoIterator<Item = Vec<Option<T>>> {
```

Comment:
Returning Vec<Option<T>> and not Result<Vec<Option<T>> will create problems with Oxigraph and streaming SPARQL evaluators: they might want to raise an error during the iteration (e.g. because of a network or disk problem).

Reply:
good point, I'll fix that

Reply:
Wouldn't it make sense to split the API between a fallible one (e.g., streaming) and an infallible one (in memory, blocking IO)?

Reply:
Segregating the API that way would reduce reusability... The path I have chosen with Sophia is to provision for errors whenever *some* implementation may fail, and use [`Infallible`](https://doc.rust-lang.org/std/convert/enum.Infallible.html) (and eventually the [`never`](https://doc.rust-lang.org/std/primitive.never.html) type) as the error of infallible implementations.

Reply:
Ok, that's reasonable. I accept :)

Reply:
That being said, you just gave me a crazy idea: https://users.rust-lang.org/t/rfc-idea-result-t-implementing-defer-t/47910